### ISSUE TITLE
Question: How to sandboxise pyodide .

### ISSUE THREAD

**Issue by Lyoko-Jeremie** [pyodide/pyodide#959]
relative to #955 , i want a sandboxed python runtime, i face two question is that :

1. how to limit the `import` function ? have any offical way todo this work ? i want limit the import with a denyList or a allowList, like the `js` i will deny it in a special `pyodide.runPythonAsync` call.
2. how to pip out the stdio ( redirect the `print` output from python ) ? now i use a way is make a patch in the `sys.stdout` from [here](https://stackoverflow.com/a/688816/3548568), but seems like it will failure after i import the `numpy` package , so strange .

---

about the question **(1)** , i was try 2 way to do. *( sorry i never read the code of the pyodide , all the follow patch work are try on runtime console and then write to code . )*
*Summary: i run the pyodide in WebWorker to use the javascript sandboxie runtime to have a additional protection .*

### 1. let the debuger hard to direct find the pyodide ref from root ( Worker's `self` or `globalThis` )

+ **in one side, i wrapper all the thing into a Typescript class , i named it `class SClass` ( mean : secure class)**

use follow code to hide it

```ts

class SClass {
    constructor() {
        workerSelf.onmessage = this.onmessage.bind(this);
    }

    // .......
```

```ts
const workerSelf: typeof WorkerSelf = self as any;
workerSelf.aaaaaa = new SClass();
workerSelf.aaaaaa.init();
// https://github.com/Microsoft/TypeScript/issues/24587#issuecomment-412287117
// we have the `workerSelf.onmessage` on the ref link, the `onmessage` are `onmessage.bind(this)`,
//   so, the GC couldn't delete our object,
//   so, we can safe delete the direct ref from root object,
//   then, we implement the hidden all the ref target,
//   no one can access our object from debug console.
// @ts-ignore
// workerSelf[Symbol() as any] = workerSelf.aaaaaa;
delete workerSelf.aaaaaa;

```

now , no one can access the SClass from `self` .

+ **in other side, because the pyodide cannot be delete from `globalThis`, (it be set the `non-configable` flag), i cloned all the field  and delete them from pyodide, only recover the `pyodide._module` to let it can use `import js`**

```ts

    init() {
        return workerSelf.languagePluginLoader.then(async () => {
            // await workerSelf.pyodide.loadPackage(['numpy', 'pytz']);

            // delete root ref from self.root
            this.pyodide = clone(workerSelf.pyodide);
            // console.log(this.pyodide);
            // console.log(this.pyodide._module);
            // console.log(Object.getOwnPropertyNames(workerSelf.pyodide));
            Object.keys(workerSelf.pyodide).forEach(T => delete workerSelf.pyodide[T]);

            // delete packages lookup-table from pyodide
            // console.log('packages', workerSelf.pyodide._module.packages);
            // console.log('packages', this.pyodide._module.packages);
            workerSelf.pyodide._module = {};
            workerSelf.pyodide._module.packages = this.pyodide._module.packages;
            workerSelf.pyodide._module.packages = {
                dependencies: {},
                import_name_to_package_name: {},
            };
            // console.log('packages', workerSelf.pyodide._module.packages);

        })
```

now, other than the `js` package cannot be import, all the try will throw error .

.

but if i only want a allowList or denyList, i cannot comlete it use follow code to recover the `pyodide._module.packages` . it's not work.

```ts

            // recover selected packages to pyodide lookup-table
            const recoverPack = (pName: string) => {
                workerSelf.pyodide._module.packages.dependencies[pName] =
                    this.pyodide._module.packages.dependencies[pName];
                workerSelf.pyodide._module.packages.import_name_to_package_name[pName] =
                    this.pyodide._module.packages.import_name_to_package_name[pName];
            };
            const allowList = [
                'numpy',
            ];
            allowList.forEach(T => recoverPack(T));
            Object.getOwnPropertyNames(this.pyodide._module.packages.dependencies).forEach(T => recoverPack(T));
            // dont remember to recover the prototype
            if (allowList.length > 0) {
                // console.log('this.pyodide._module.packages.prototype',
                //     Object.getPrototypeOf(this.pyodide._module.packages));
                Object.setPrototypeOf(workerSelf.pyodide._module.packages,
                    Object.getPrototypeOf(this.pyodide._module.packages));
                Object.setPrototypeOf(workerSelf.pyodide._module.packages.dependencies,
                    Object.getPrototypeOf(this.pyodide._module.packages.dependencies));
                Object.setPrototypeOf(workerSelf.pyodide._module.packages.import_name_to_package_name,
                    Object.getPrototypeOf(this.pyodide._module.packages.import_name_to_package_name));
            }
            console.log('packages', workerSelf.pyodide._module.packages);
```

seems like some other important thing i deleted from the `workerSelf.pyodide`.

in the end, i remove above the attempt that delete thing from pyodide.

### 2. use `pyodide.find_imports` to check the code before run it . and simple not run it if find something not allowed.

[see pyodide.find_imports](https://pyodide.readthedocs.io/en/latest/python-api/pyodide.find_imports.html#pyodide.find_imports)

thsi is a easy way, but [in the document](https://pyodide.readthedocs.io/en/latest/api_reference.html#python-api) i see some dangerous function can bypass this way. like `pyodide.eval_code` or the `pyodide.open_url` , and i not sure not have any other way can load code from remote or eval a string to code in runtime.

---

BTW: the `cursor` params of `pyodide.get_completions` [API](https://pyodide.readthedocs.io/en/latest/python-api/pyodide.get_completions.html#pyodide.get_completions) seems like not receive {row, col} as params, i dont known how to use it in a large section code. so, now i run a second WebWorker to run the `Jedi` to do the code complete task. but seems like the `jedi` cannot get information from a imported package , like `numpy` .

*sorry for my knowledge, i'm good at C++/Typescript and software security/software architecture, the Python is in my weakness area.*

**Comment by hoodmane** [comment]
Pyodide comes sandboxed. Pyodide runs in the browser. The browser already sandboxes each browser tab running inside of it. So yeah, just do nothing, it's sandboxed out of the box.

**Comment by hoodmane** [comment]
> how to pip out the stdio ( redirect the print output from python ) ? now i use a way is make a patch in the sys.stdout from here, but seems like it will failure after i import the numpy package , so strange .

Could you please open a separate issue about that with more info? It sounds like a bug.

**Comment by hoodmane** [comment]
> the cursor params of `pyodide.get_completions` API seems like not receive {row, col} as params

The documentation for that is very vague.  I suggest not using this API. Instead:
```javascript
let completions = await pyodide.runPythonAsync(`
   import jedi
   interpreter = jedi.Interpreter(code, [namespace])
   jedi_signatures =  interpreter.complete(line=lineNumber, column=column, fuzz=fuzzy)
`)
```
[See the Jedi docs here.](https://jedi.readthedocs.io/en/latest/docs/api.html#jedi.Script.complete).

Note that Jedi in pyodide has some fundamental limitations, particularly in Chrome, which cannot be resolved without changes to the browser (see #821).

**Comment by Lyoko-Jeremie** [comment]
> Pyodide comes sandboxed

I think that not true.

If a Python script can use `import js` , then it can use `js.postMessage('{"badKey": "badValue"}')` on WebWorker or `js.whatEverOther.EvilThing` to chang the js env that the pyodide it self working on. 

Example, the script can send HTTP request or modify a js object, and it can call the 'js.eval' function to covert some code string to js code to inject into js runtime that the code string can be download from remote server.

This above example is a classic Remote Code Execute issue in a sandboxes .

And even though i filter the `import js` statement ,  i cannot sure don't have any something others way can access the js object , this is a typical sandboxie escape issue .

So , in my opinion , pyodido not a good sandboxie for my case.

**Comment by hoodmane** [comment]
There is no point in sandboxing pyodide: an "evil user" can always open the browser console and type arbitrary code into that. If you are planning to make a server that will communicate with the pyodide session, the server needs to treat all information from the client as untrusted. No matter what your server needs to be secure and you cannot make the client anymore secure than it is out of the box.

**Comment by dalcde** [comment]
I can envision several use cases where further sandboxing may be
desired.

1. To prevent XSS attacks in notebook-like environment, where python
    code may be executed in the browser once a page is opened.

2. When one wants to use wasm as a sandbox to run user python code
    server-side

**Comment by hoodmane** [comment]
> 2. When one wants to use wasm as a sandbox to run user python code server-side

But then wasm is the sandbox and pyodide doesn't need changes? [It says here](https://wiki.python.org/moin/SandboxedPython) that sandboxing the CPython vm is hard. Do you think it's easier to sandbox the pyodide CPython than the ordinary one?

**Comment by dalcde** [comment]
One might want to prevent pyodide from making network requests, for
example. By default, the wasm sandbox is only allowed to use explicitly
imported JS functions, and you can't make network requests from within
the wasm sandbox directly.

**Comment by hoodmane** [comment]
I see, thanks for explaining. Then there is something reasonable that could be done. I guess the conclusion is that this is a feature request.

**Comment by dalcde** [comment]
I think it would be reasonable to add a compile time option to disable
all networking functionality.

**Comment by Lyoko-Jeremie** [comment]
@dalcde i think the simple use way is make a configure to let pyodide runner can special what js object can be access from Python side , 

Example, provide a config to disable the js feature in a `runPythonAsync` call .

Or, make a config to let caller to fill what thing is the `js` object , not always the hole `globalThis` .

**Comment by hoodmane** [comment]
> Or, make a config to let caller to fill what thing is the js object , not always the hole globalThis .

That is a good idea. We can add a `pyodide.mountPackage` api or something like that that adds a javascript object that can be imported from. I will add it to my todo list.

**Comment by hoodmane** [comment]
@dalcde To be honest, I'm still a bit confused though about the sandboxing issue. Who are you trying to protect from what by blocking pyodide from making network requests? It must be the user because the server can't be protected from receiving random network requests. Is the idea that the user is less likely to paste random stuff into their browser console than into the a pyodide console?

**Comment by Lyoko-Jeremie** [comment]
> an "evil user" can always open the browser console and type window.whatEverOther.EvilThing into that. 

About this , you can see how i hidden my `SClass` in the top . 

Use that way to hidden the SClass , every one include me cannot access the SClass instance from the Browser console . I checked this in Firefox last night. And will to check this in Chrome if i need.

**Comment by hoodmane** [comment]
But what could they do if they could access `SClass` that you are trying to prevent?

**Comment by hoodmane** [comment]
The user could use the browser dev tools to edit the javascript source and then access `SClass`.

**Comment by dalcde** [comment]
On Sat, Dec 26, 2020 at 06:21:49PM -0800, Hood Chatham wrote:
>@dalcde To be honest, I'm still a bit confused though about the sandboxing issue. Who are you trying to protect from what by blocking pyodide from making network requests? It must be the user because the server can't be protected from receiving random network requests. Is the idea that the user is less likely to paste random stuff into their browser console than into the a pyodide console?

I'm considering the situation where pyodide is *not* run in a browser, but in some server-side environment. If you want to run untrusted python code from users (e.g. a coding challenge), it could be reasonable to try to run it in a wasm-sandboxed python runtime.

In this case, there are many good reasons to not want the user to be able to make network requests. For example, users might use your service to launch malicious attacks of other (unrelated) services. This could also let users access your internal network, which may be undesirable.

**Comment by oeway** [comment]
Sandboxing is really hard, I doubt it worth the effort. For example, to access network, one can also compile a python module by using the fetch api: https://emscripten.org/docs/api_reference/fetch.html


> 1. To prevent XSS attacks in notebook-like environment, where python code may be executed in the browser once a page is opened.

Since jsfiddle, codepen, etc. are similar to a notebook and already allowing executing untrusted script once the page is opened, maybe we don't really need to prevent XSS attack from pyodide itself?

**Comment by hoodmane** [comment]
@dalcde Okay I understand what you are saying now.
@oeway It says right there though that the Fetch API is opt-in. I think the idea is that to sandbox you'd opt out of these things:
> In order to use the Fetch API, you would need to compile your code with -s FETCH=1.

**Comment by Lyoko-Jeremie** [comment]
> If you want to run untrusted python code from users (e.g. a coding challenge)

Yes , i writing a Online Juge like software , you can think it is similar like a CTF , the code run in pyodide are come from the user, that user can do anything if you not deny it. 

As the Chinese says goes , You can do things that are not prohibited by the law.

**Comment by hoodmane** [comment]
@Lyoko-Jeremie I'm really confused. @dalcde is talking about running code on a server, where there is a clear point to limiting the rights of the code. I think you are talking about running code on the browser, where I don't think there is a point to it.

**Comment by Lyoko-Jeremie** [comment]
I plane to split the SClass code to multiple part use crypto function , and encryption every part with asymmetric encryption , then decrypt  and  merge before use , then use eval() function to load then , so , a evil user cannot simply put breakpoint on SClass or access it's memory.

**Comment by Lyoko-Jeremie** [comment]
@hoodmane i think the server and client have similar attacks model

**Comment by hoodmane** [comment]
@Lyoko-Jeremie You still haven't explained what special thing `SClass` can do that can't be done without access to it. 

>a evil user cannot simply put breakpoint on SClass or access it's memory.

You cannot prevent an evil user from accessing the memory of `SClass`. What if they are running a debug build of the browser and just manually inspect the memory? Any memory on the user's computer cannot be impossible to access, only annoying to access.

**Comment by Lyoko-Jeremie** [comment]
> Any memory on the user's computer cannot be impossible to access, only annoying to access.

Yes , so i not want to make a perfect security env . 

Like the AES algorithm OR any other software protect system OR other anything we are using in cryptography , nothing is perfect security , i only want to let the evil user must make more and more and more and more proof to inspect the software .

**Comment by Lyoko-Jeremie** [comment]
> You still haven't explained what special thing SClass can do that can't be done without access to it.

I will run all the thing which need protect in the SClass , so i named it `secure class`

**Comment by Lyoko-Jeremie** [comment]
Because of some reason , the code challenge cannot running in server side , so i make the best move , like a online fighting game engineering do on anti cheats .

How frustrating .

**Comment by hoodmane** [comment]
> code challenge ... anti cheats 

Oh okay, now I understand. Thanks for explaining. What you are describing sounds entirely sufficient: your system sounds annoying enough to break into that if anyone _does_ hack the system, they deserve to win your code challenge.

**Comment by dalcde** [comment]
On Sat, Dec 26, 2020 at 06:47:39PM -0800, Wei Ouyang wrote:
>Sandboxing is really hard, I doubt it worth the effort. For example, to access network, one can also compile a python module by using the fetch api: https://emscripten.org/docs/api_reference/fetch.html

For this to work, I believe you need to set -s FETCH=1 when building the
main module, because again by default the fetch function is not
available inside wasm unless we explicitly import it when initiating the
wasm module from javascript.

For us I believe we mostly cross the wasm-js boundary through lots of
EM_JS.

>> 1. To prevent XSS attacks in notebook-like environment, where python code may be executed in the browser once a page is opened.
>
>Since jsfiddle, codepen, etc. are similar to a notebook and already allowing executing untrusted script once the page is opened, maybe we don't really need to prevent XSS attack from pyodide itself?

It is of course possible to prevent XSS via other means, but depending
on your use case that may or may not be more troublesome than doing it
through pyodide itself (and if course, you might actually want to let
people make network requests in your notebook, in which case you have no
choice but to do protection via other means).

**Comment by dalcde** [comment]
I should add that if you want to run pyodide in node, then having full JS access would be a bad idea for a different reason.

**Comment by Lyoko-Jeremie** [comment]
@hoodmane 
Yes , but not only this ,

To Be a example, my project will using in a competition , if some one hack it and no one perceive , the hacker will direct enter the MIT without SAT . But if someone was found cheating , the hacker's school life is over , and will be banned on any exam in 3~5 years by the national power .

**Comment by hoodmane** [comment]
> banned on any exam in 3~5 years by the national power .

Sounds scary.

**Comment by hoodmane** [comment]
Okay this is a long thread. I opened #960. Is there anything else from this that should be implemented?

**Comment by Lyoko-Jeremie** [comment]
> Sounds scary.

This is necessary for fairness, isn't it?

**Comment by Lyoko-Jeremie** [comment]
*and BTW: `banned on any exam in 3~5 years by the national power` This Really Happend and This Line Has Been Written Into Law.*

My job is to be as fair as possible, and try to prevent cheat . Because the champion will win the Examination-free Admission and who can select any school if he/she want to go. **it is real.**

**Comment by hoodmane** [comment]
Closed as #960 was implemented. This is a long thread, if there are other desired features here please open a new issue.
